src/Command.ts

```diff
@@ -169,14 +172,50 @@ export class Command {
 		return fulfillsRequirements(this.requirements, msg, args, ctx);
 	}
 
+	/**
+	 * Adds a subcommand to this command.
+	 * @param command The subcommand to add
+	 */
+	addSubcommand (command: Command): this {
+		for (const name of command.names) {
+			for (const otherCommand of this.subcommands) {
+				if (otherCommand.names.includes(name)) {
+					throw new TypeError(`Two commands have the same name: ${name}`);
+				}
+			}
+		}
+		this.subcommands.push(command);
+		return this;
+	}
+
+	/**
+	 * Checks the list of subcommands and returns one whch is known by a given
+	 * name. Passing an empty string will return the default command, if any.
+	 */
+	subcommandForName (name: string, caseSensitive: boolean): Command | null {
+		if (caseSensitive) return this.subcommands.find(c => c.names.includes(name)) || null;
+		return this.subcommands.find(c => c.names.some(n => n.toLowerCase() === name.toLowerCase())) || null;
+	}
+
 	/** Executes the command process if the permission checks pass. */
 	async execute(msg: Eris.Message, args: string[], ctx: CommandContext): Promise<boolean> {
 		if (!await this.checkPermissions(msg, args, ctx)) return false;
-		// By calling checkPermissions and returning early if it returns false,
-		// we guarantee that messages will be the correct type for the stored
-		// process, so this call is always safe. Restructuring this to properly
-		// use TS type guards would be very messy and would result in duplicate
-		// safety checks that we want to avoid.
+
+		// Check if we have a subcommand, and if so, execute that command
+		const subcommand = this.subcommandForName(args[0], ctx.client.caseSensitiveCommands);
+		if (subcommand) {
+			// TODO: Might want to handle this as an array instead, but doing it
+			//       this way for now for backwards compatibility
+			ctx.commandName += ` ${args.shift()}`;
+			return subcommand.execute(msg, args, ctx);
+		}
```

Comment:
Eight mentioned that they were getting an `undefined` exception when executing a command while giving no arguments, I believe this might be the problem. I could however be completely barking up the wrong tree. Image of exception [here](https://cdn.discordapp.com/attachments/409840359632404481/829659353841008690/unknown.png)
```suggestion
		// Check if args is defined.
		if(args) {
			// Check if we have a subcommand, and if so, execute that command
			const subcommand = this.subcommandForName(args[0], ctx.client.caseSensitiveCommands);
			if (subcommand) {
				// TODO: Might want to handle this as an array instead, but doing it
				//       this way for now for backwards compatibility
				ctx.commandName += ` ${args.shift()}`;
				return subcommand.execute(msg, args, ctx);
			}
		}
```

Reply:
Done with 5231ecc, thanks again for investigating!